staging/vhost-device-gpu/src/device.rs

```diff
@@ -1385,4 +1390,227 @@ mod tests {
+
+        backend.set_gpu_socket(gpu_backend);
+
+        // Unfortunately there is no way to crate a VringEpollHandler directly (the ::new is not public)
+        // So we create a daemon to create the epoll handler for us here
+        let daemon = VhostUserDaemon::new(
+            "vhost-device-gpu-backend".to_string(),
+            backend.clone(),
+            mem.clone(),
+        )
+        .expect("Could not create daemon");
+        let epoll_handlers = daemon.get_epoll_handlers();
+        backend.set_epoll_handler(&epoll_handlers);
+        mem::drop(daemon);
+
+        const IMAGE_ADDR: GuestAddress = GuestAddress(0x30_000);
+        const IMAGE_WIDTH: u32 = 640;
+        const IMAGE_HEIGHT: u32 = 480;
+        const RESP_SIZE: u32 = mem::size_of::<virtio_gpu_ctrl_hdr>() as u32;
+
+        let image_rect = virtio_gpu_rect {
+            x: 0,
+            y: 0,
+            width: IMAGE_WIDTH,
+            height: IMAGE_HEIGHT,
+        };
+
+        let hdr = new_hdr(VIRTIO_GPU_CMD_RESOURCE_CREATE_2D);
+        let cmd = virtio_gpu_resource_create_2d {
+            resource_id: 1,
+            format: VIRTIO_GPU_FORMAT_R8G8B8A8_UNORM, // RGBA8888
+            width: IMAGE_WIDTH,
+            height: IMAGE_HEIGHT,
+        };
+        let create_resource_cmd = TestingDescChainArgs {
+            readable_desc_bufs: &[hdr.as_slice(), cmd.as_slice()],
+            writable_desc_lengths: &[RESP_SIZE],
+        };
+
+        let hdr = new_hdr(VIRTIO_GPU_CMD_RESOURCE_ATTACH_BACKING);
+        let mem_entries = split_into_mem_entries(IMAGE_ADDR, IMAGE_WIDTH * IMAGE_HEIGHT * 4, 4096);
+
+        let cmd = virtio_gpu_resource_attach_backing {
+            resource_id: 1,
+            nr_entries: mem_entries.len() as u32,
+        };
+        let mut readable_desc_bufs = vec![hdr.as_slice(), cmd.as_slice()];
+        readable_desc_bufs.extend(mem_entries.iter().map(|entry| entry.as_slice()));
+        let attach_backing_cmd = TestingDescChainArgs {
+            readable_desc_bufs: &readable_desc_bufs,
+            writable_desc_lengths: &[RESP_SIZE],
+        };
+
+        let hdr = new_hdr(VIRTIO_GPU_CMD_TRANSFER_TO_HOST_2D);
+        let cmd = virtio_gpu_transfer_to_host_2d {
+            r: image_rect,
+            offset: 0,
+            resource_id: 1,
+            padding: Default::default(),
+        };
+        let transfer_to_host_cmd = TestingDescChainArgs {
+            readable_desc_bufs: &[hdr.as_slice(), cmd.as_slice()],
+            writable_desc_lengths: &[RESP_SIZE],
+        };
+
+        let hdr = new_hdr(VIRTIO_GPU_CMD_SET_SCANOUT);
+        let cmd = virtio_gpu_set_scanout {
+            r: image_rect,
+            resource_id: 1,
+            scanout_id: 1,
+        };
+        let set_scanout_cmd = TestingDescChainArgs {
+            readable_desc_bufs: &[hdr.as_slice(), cmd.as_slice()],
+            writable_desc_lengths: &[RESP_SIZE],
+        };
+
+        let hdr = new_hdr(VIRTIO_GPU_CMD_RESOURCE_FLUSH);
```

Comment:
I think you can add more comments in this commit e.g `// Flush resource command`